Set correct device id on efficientzerotensors

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71611

Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925 #69913

Differential Revision: [D33897543](https://our.internmc.facebook.com/intern/diff/D33897543)

anjali411 committed:
Set correct device id on efficientzerotensors

[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"




[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"




[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"




[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"




[ghstack-poisoned]

albanD commented on c10/core/TensorImpl.h:
```diff
@@ -1092,9 +1092,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
-      TORCH_INTERNAL_ASSERT(
-          false,
-          "Please call `torch._efficientzerotensor` if you want to create a tensor with no storage.");
+      key_set_ = key_set_.add(DispatchKey::ZeroTensor);
```
If you don't want to change this, you can also add this key to the DispatchKeySet you create in the factory functions. Why relax this here?

anjali411 replied:
sure

albanD commented on test/test_torch.py:
```diff
@@ -5475,6 +5475,13 @@ def test_clamp(self, device, dtype):
             actual = x[..., :1].clamp(lb, ub)
             self.assertEqual(expect, actual)
 
+    @onlyCUDA
```
Why not run this on all other devices?

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33768645).

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925

Differential Revision: [D33768645](https://our.internmc.facebook.com/intern/diff/D33768645)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33768645).

mruberry commented:
Unlanding. This appears to be breaking several jobs (or exacerbating existing failures so they're happening consistently). Some of the failures appear in PR CI: 

```
======================================================================
ERROR [0.030s]: test_fn_fwgrad_bwgrad_nn_functional_dropout_cuda_float64 (__main__.TestGradientsCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1580, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 753, in test_wrapper
    return test(*args, **kwargs)
  File "test_ops.py", line 827, in test_fn_fwgrad_bwgrad
    self._check_helper(device, dtype, op, op.get_op(), "fwgrad_bwgrad")
  File "test_ops.py", line 777, in _check_helper
    self.assertTrue(gradgradcheck(fn, gradcheck_args, **kwargs))
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2956, in gradgradcheck
    return torch.autograd.gradgradcheck(fn, inputs, grad_outputs, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1550, in gradgradcheck
    check_forward_ad=check_fwd_over_rev, check_backward_ad=check_rev_over_rev)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1400, in gradcheck
    return _gradcheck_helper(**args)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1417, in _gradcheck_helper
    check_undefined_grad=check_undefined_grad)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1087, in _gradcheck_real_imag
    rtol, atol, check_grad_dtypes, nondet_tol, use_forward_ad=True)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1286, in _fast_gradcheck
    all_u=all_u, check_grad_dtypes=check_grad_dtypes)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 343, in _get_analytical_jacobian_forward_ad
    raw_outputs = _as_tuple(fn(*dual_inputs))
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1538, in new_func
    outputs = _differentiable_outputs(func(*input_args))
  File "test_ops.py", line 740, in fn
    output = op.gradcheck_wrapper(variant, *inputs, **sample.kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_methods_invocations.py", line 8131, in wrapper_set_seed
    return op(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_methods_invocations.py", line 14546, in <lambda>
    wrapper_set_seed(torch.nn.functional.dropout, input, *args, **kwargs),
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_methods_invocations.py", line 8131, in wrapper_set_seed
    return op(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/functional.py", line 1279, in dropout
    return _VF.dropout_(input, p, training) if inplace else _VF.dropout(input, p, training)
NotImplementedError: Trying to use forward AD with native_dropout that does not support it.

======================================================================
ERROR [0.157s]: test_forward_mode_AD_nn_functional_batch_norm_cuda_float64 (__main__.TestGradientsCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1580, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
         4.1013e-02, -3.1660e-02, -2.3235e-02, -8.5606e-02,  9.9489e-02,
        -2.9030e-02,  1.1628e-02,  1.2536e-02,  1.5776e-02, -1.0911e-02,
        -8.8994e-02,  6.6432e-02,  3.0821e-02,  7.0483e-02, -7.8742e-02,
         1.1666e-01, -1.3688e-01, -1.1689e-01,  2.1548e-02,  1.1556e-01,
         2.0140e-03, -4.7531e-03, -1.0538e-04,  3.0550e-03, -2.1054e-04],
       device='cuda:0', dtype=torch.float64)
analytical:tensor([  2.8588,  -2.0172,   2.3206,   4.5918,  -7.7540,   0.0388,  -1.5723,
          6.6986,  -2.5357,  -2.6295,   0.6157,  -1.8667,   1.5832,  -2.5946,
          2.2624,   0.7529,   0.6592,   1.9918,  -0.8375,  -2.5665,  -0.4558,
          0.5823,  -0.3433,   0.7596,  -0.5429,   0.0390,  -0.5325,   0.0970,
          0.5532,  -0.1568,   2.6474,  -2.0865,   1.6848,  -1.3425,  -0.9033,
          1.1633,   2.0510,   0.1307,   1.5361,  -4.8811,  -0.1976,   0.2526,
         -0.0502,  -0.0345,   0.0296,  -2.5113,   3.9516,  -4.6885,   0.2703,
          2.9779,   6.5899,  -3.8758,  -1.2627,  -4.7210,   3.2695,   0.8214,
          0.3628,   0.2740,  -0.8045,  -0.6536,  -1.0287,  -0.5708,  -0.8996,
          0.4038,   2.0953,  -0.3568,   0.8307,   0.1297,  -0.2995,  -0.3041,
         -0.5026,   0.4587,   0.3865,   0.6740,  -1.0165,   2.4162,   0.5919,
          0.2770,  -2.8751,  -0.4100,   0.1767,  -0.0975,  -0.4603,   0.3252,
          0.0559, -12.4702,   5.4662,   5.0397,  -5.6636,   7.6280,   3.5727,
         -1.4230,   0.9862,  -1.9007,  -1.2353,  -0.9009,  -1.1699,   0.6029,
          1.3876,   0.0803,  -0.5537,   1.0734,  -2.3366,   0.4858,   1.3312,
         -2.4204,  -2.1454,   1.6810,  -1.4707,   4.3556,  -4.5250,  -4.5533,
          4.6526,   1.5535,   2.8721,   0.1622,  -0.2986,  -0.3440,   0.0754,
          0.4051,  -3.8046,   0.7902,  -4.0687,   3.5868,   3.4963],
       device='cuda:0', dtype=torch.float64, grad_fn=<CopyBackwards>)

The above quantities relating the numerical and analytical jacobians are computed 
in fast mode. See: https://github.com/pytorch/pytorch/issues/53876 for more background 
about fast mode. Below, we recompute numerical and analytical jacobians in slow mode:

Numerical:
 tensor([[ 1.0669, -0.1998, -0.4058,  ...,  0.0000,  0.0000,  0.0000],
        [-0.1998,  1.1230, -0.2192,  ...,  0.0000,  0.0000,  0.0000],
        [-0.4058, -0.2192,  1.1132,  ...,  0.0000,  0.0000,  0.0000],
        ...,
        [ 0.0000,  0.0000,  0.0000,  ...,  0.1605,  0.0164,  0.0150],
        [ 0.0000,  0.0000,  0.0000,  ...,  0.0164,  0.1817, -0.1308],
        [ 0.0000,  0.0000,  0.0000,  ...,  0.0150, -0.1308,  0.1841]],
       device='cuda:0', dtype=torch.float64)
Analytical:
tensor([[ -7.6410,   8.6218, -12.0329,  ...,   0.0000,   0.0000,   0.0000],
        [  8.6218,  -2.0228,   6.6744,  ...,   0.0000,   0.0000,   0.0000],
        [-12.0329,   6.6744,  -3.0030,  ...,   0.0000,   0.0000,   0.0000],
        ...,
        [  0.0000,   0.0000,   0.0000,  ..., -53.2109,  47.9129,  47.0579],
        [  0.0000,   0.0000,   0.0000,  ...,  47.9129, -40.2781, -42.1931],
        [  0.0000,   0.0000,   0.0000,  ...,  47.0579, -42.1931, -38.7944]],
       device='cuda:0', dtype=torch.float64, grad_fn=<CopySlices>)

The max per-element difference (slow mode) is: 162.0326729297527.


----------------------------------------------------------------------
Ran 13709 tests in 3355.462s

FAILED (errors=3, skipped=5765, expected failures=97, unexpected successes=2)
```

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925


[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925


[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925


[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33834916).

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925

Differential Revision: [D33834916](https://our.internmc.facebook.com/intern/diff/D33834916)

[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925

Differential Revision: [D33834916](https://our.internmc.facebook.com/intern/diff/D33834916)

[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925

Differential Revision: [D33834916](https://our.internmc.facebook.com/intern/diff/D33834916)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33834916).

soulitzer commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9069,9 +9069,6 @@ def ref_pairwise_distance(input1, input2):
-                        # 69913: RuntimeError: CUDA error: an illegal memory access was encountered
-                        DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_fn_fwgrad_bwgrad',
-                                     device_type='cuda', dtypes=[torch.double, torch.cdouble]),
                         DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_forward_mode_AD',
```
After patching this PR,  plain forward mode AD tests also pass. We should remove those skips as well.

soulitzer replied:
Could you also double check for the inplace ones?

anjali411 replied:
I'll submit a follow-up PR to remove more skips. First let's get this fix in after rocm test passes

anjali411 commented:
unlanding because `inux-bionic-cuda10.2-py3.9-gcc7 / test (default, 2, 2, linux.4xlarge.nvidia.gpu)` is failing on master:

```
======================================================================
ERROR [0.242s]: test_fn_grad_linalg_det_singular_cuda_float64 (__main__.TestGradientsCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1753, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 381, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 788, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 753, in test_wrapper
    return test(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_ops.py", line 834, in test_fn_grad
    self._grad_test_helper(device, dtype, op, op.get_op())
  File "/var/lib/jenkins/workspace/test/test_ops.py", line 820, in _grad_test_helper
    return self._check_helper(device, dtype, op, variant, 'gradcheck', check_forward_ad=check_forward_ad,
  File "/var/lib/jenkins/workspace/test/test_ops.py", line 789, in _check_helper
    self.assertTrue(gradcheck(fn, gradcheck_args,
  File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 2890, in gradcheck
    return torch.autograd.gradcheck(fn, inputs, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/gradcheck.py", line 1400, in gradcheck
    return _gradcheck_helper(**args)
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/gradcheck.py", line 1414, in _gradcheck_helper
    _gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps,
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/gradcheck.py", line 1061, in _gradcheck_real_imag
    gradcheck_fn(func, func_out, tupled_inputs, outputs, eps,
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/gradcheck.py", line 1291, in _fast_gradcheck
    analytical_vJu = _get_analytical_vJu_backward_mode(inputs, outputs, nondet_tol, check_grad_dtypes, all_v, all_u_dense)
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/gradcheck.py", line 555, in _get_analytical_vJu_backward_mode
    all_vJ = _check_analytical_jacobian_attributes(inputs, output, nondet_tol, check_grad_dtypes,
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/gradcheck.py", line 529, in _check_analytical_jacobian_attributes
    vjps2 = _get_analytical_vjps_wrt_specific_output(vjp_fn, output.clone(), v)
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/gradcheck.py", line 633, in _get_analytical_vjps_wrt_specific_output
    grad_inputs = vjp_fn(v.reshape(sample_output.shape))
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/gradcheck.py", line 524, in vjp_fn
    return torch.autograd.grad(output, diff_input_list, grad_output,
  File "/opt/conda/lib/python3.9/site-packages/torch/autograd/__init__.py", line 275, in grad
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.

----------------------------------------------------------------------
Ran 7649 tests in 635.801s
```

malfet commented:
@anjali411 do you know why CI testing didn't caught that? Or had we added more tests?

anjali411 commented:
> @anjali411 do you know why CI testing didn't caught that? Or had we added more tests?

I didn't add any more tests.. I am not sure but it looks like this is a persistent issue on master and was just recently disabled https://github.com/pytorch/pytorch/issues/71808#issuecomment-1026043325

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33897543).

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925 #69913

Differential Revision: [D33897543](https://our.internmc.facebook.com/intern/diff/D33897543)

[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925 #69913

Differential Revision: [D33897543](https://our.internmc.facebook.com/intern/diff/D33897543)

[ghstack-poisoned]

anjali411 committed:
Update on "Set correct device id on efficientzerotensors"


Fixes https://github.com/pytorch/pytorch/issues/71160 https://github.com/pytorch/pytorch/issues/69925 #69913

Differential Revision: [D33897543](https://our.internmc.facebook.com/intern/diff/D33897543)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33897543).

anjali411 commented:
was merged before the 1.11 branch cut
